msal_extensions/token_cache.py

```diff
@@ -0,0 +1,196 @@
+
+        self._cache_location = os.path.expanduser(self._cache_location)
+        self._lock_location = os.path.expanduser(self._lock_location)
+
+        _mkdir_p(os.path.dirname(self._lock_location))
+        _mkdir_p(os.path.dirname(self._cache_location))
+
+    def _needs_refresh(self):
+        # type: () -> Bool
+        """
+        Inspects the file holding the encrypted TokenCache to see if a read is necessary.
+        :return: True if there are changes not reflected in memory, False otherwise.
+        """
+        try:
+            return self._last_sync < os.path.getmtime(self._cache_location)
+        except IOError as exp:
+            if exp.errno != errno.ENOENT:
+                raise exp
+            return False
+
+    def _write(self):
+        """Handles actually committing the serialized form of this TokenCache to persisted storage.
+        For types derived of this, class that will be a file, which has the ability to track a last
+        modified time.
+        """
+        with open(self._cache_location, 'wb') as handle:
+            handle.write(self.serialize())
+
+    def _read(self):
+        """Fetches the contents of a file and invokes deserialization."""
+        with open(self._cache_location, 'rb') as handle:
+            contents = handle.read()
+        self.deserialize(contents)
+
+    def add(self, event, **kwargs):
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            super(FileTokenCache, self).add(event, **kwargs)  # pylint: disable=duplicate-code
+            self._write()
+
+    def modify(self, credential_type, old_entry, new_key_value_pairs=None):
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            super(FileTokenCache, self).modify(
+                credential_type,
+                old_entry,
+                new_key_value_pairs=new_key_value_pairs)
+            self._write()
+
+    def find(self, credential_type, **kwargs):  # pylint: disable=arguments-differ
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            return super(FileTokenCache, self).find(credential_type, **kwargs)
+
+    def __getattr__(self, item):
+        # Instead of relying on implementers remembering to update _last_sync, just detect that one
+        # of the relevant methods has been called take care of it for derived types.
```

Comment:
Thank you for paying attention to this kind of details, and coming up with a creative solution!

On the other hand, how about consider NOT do this `__getattr__(...)` magic (which even pylint complains for), and do an old-school inheritance instead?

```python
def serialize():
    self._last_sync = int(time.time())  # In fact, this is the more precise spot to do it, i.e. this way you won't update the time if an IOError would happen
    return super(..., self).serialize()
    # And similar approach to deserialize()
```

Reply:
I've been harboring some reservations about this magic myself! I like the idea of keeping `_last_sync` out of the `_read` and `_write` implementations altogether. However, I'd like to keep `serialize` and `deserialize` as free of side-effects as possible. For that reason I'll move it into `modify`, `find`, and `add` which innately have side effects! The other thing I realized over the last couple of days, is that it's a lot more reliable to use the last modified time from the file rather than capturing that time via the Python system timer. While I have the hold on the CrossPlatLock, I'll read that metadata and use that for the value of `_last_sync`

Reply:
I really appreciate your conscious effort on keeping the `serialize()` and `deserialize()` free of side effect. [Their current status](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/0.4.0/msal/token_cache.py#L292-L304) is indeed [as simple as possible, and that wasn't easy](https://www.azquotes.com/picture-quotes/quote-there-are-two-ways-of-constructing-a-software-design-one-way-is-to-make-it-so-simple-tony-hoare-54-50-35.jpg). That being said, if we consider the "last_sync" time as an intrinsic characteristic of the serialization/deserialization itself, it is OK to record it there, by just one extra line of code.
But again, we are open if you have different ideas. :-)

Reply:
Fixed in 1dff899c2c552e7c9b9442432fbeac981b488fac